### PR TITLE
feat: squad post moderation page

### PR DESCRIPTION
**Pull request by sshanzel**
## Changes
- Squad moderation page.
  - Header tabs.
  - Default squad header to avoid layout shift.
- Utilized the existing `ReportModal` for selecting a reason to reject the post.
- Squad moderation list including the button to approve all requests.
- Squad moderation item.
- Hooks to manage processing streamlined within the apps and be reused anywhere else.
- After taking on the primary task (list), I didn't realize how relative it was to some other tickets, and hence this clears 3 tickets at once 😅 

Preview:

Screenshot 2024-10-06 at 2 44 09 AM

## Events

Did you introduce any new tracking events?

## Experiment

Did you introduce any new experiments?

## Manual Testing

> [!CAUTION]
> Please make sure existing components are not breaking/affected by this PR

MI-572 #done
MI-573 #done
MI-574 #done

### Preview domain
https://mi-572.preview.app.daily.dev

### CODE REVIEW

File: packages/shared/src/hooks/squads/useSquadPostModeration.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  },
+  { value: PostModerationReason.Other, label: 'Other' },
+];
+
+interface UseSquadPostModeration {
+  onApprove: (ids: string[]) => Promise<void>;
+  onReject: (id: string) => void;
+  isLoading: boolean;
+  isSuccess: boolean;
+}
```
Comment (sshanzel) [inline packages/shared/src/hooks/squads/useSquadPostModeration.ts:53]: This is basically the `CommonModerationProps` from the RFC, just had to rename the name to follow the hook's naming convention.

File: packages/shared/src/hooks/squads/useSquadPendingPosts.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  SourcePostModeration,
+  SourcePostModerationStatus,
+} from '../../graphql/squads';
+
+interface UseSquadPendingPosts {
+  data: SourcePostModeration[];
+  isFetched: boolean;
+}
+
+export const useSquadPendingPosts = (): UseSquadPendingPosts => {
```
Comment (sshanzel) [inline packages/shared/src/hooks/squads/useSquadPendingPosts.ts:12]: To be worked on once the API query is ready.

File: packages/shared/src/graphql/squads.ts [hunk 3]
```diff
@@ -555,3 +555,46 @@ export const getPublicSquadRequests = async (
...
+}
+
+export const squadApproveMutation = (
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  _: string[],
+): Promise<void> => Promise.resolve();
+export const squadRejectMutation = (
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  _: SquadPostRejectionProps,
+): Promise<void> => Promise.resolve();
```
Comment (sshanzel) [inline packages/shared/src/graphql/squads.ts:600]: To be worked on once the API mutations are ready.

File: packages/shared/src/components/squads/moderation/SquadModerationList.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+export function SquadModerationList(): ReactElement {
+  const { onApprove, onReject, isLoading } = useSquadPostModeration();
+  const { data, isFetched } = useSquadPendingPosts();
+
+  if (!data?.length) {
+    if (!isFetched) {
+      return (
+        <div className="flex flex-col gap-4 p-6">
+          <span className="typo-title3">Loading...</span>
```
Comment (sshanzel) [inline packages/shared/src/components/squads/moderation/SquadModerationList.tsx:23]: We don't have yet a skeleton, but maybe we can come up with a simple one.

File: packages/shared/src/components/squads/moderation/SquadModerationItem.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+export function SquadModerationItem({
+  data,
+  onReject,
+  onApprove,
+  isLoading,
+}: SquadModerationListProps): ReactElement {
+  const { post } = data;
+
+  return (
+    <div className="flex flex-col gap-4 p-6">
```
Comment (sshanzel) [inline packages/shared/src/components/squads/moderation/SquadModerationItem.tsx:30]: The intricacies of which status to display based on the user's role is going to be raised in a separate PR. From this ticket: https://dailydotdev.atlassian.net/jira/software/projects/MI/boards/31?selectedIssue=MI-573

File: packages/shared/src/components/layout/common.tsx [hunk 6]
```diff
@@ -163,3 +163,10 @@ export const SearchControlHeader = ({
     </ConditionalWrapper>
   );
 };
+
+export const PageHeader = classed(
+  'div',
+  'flex flex-row items-center border-b border-border-subtlest-tertiary px-4 py-2',
+);
+
+export const PageHeaderTitle = classed('h2', 'px-4 font-bold');
```
Comment (rebelchris) [inline packages/shared/src/components/layout/common.tsx:172]: non blocking, you could have also use `Typography` instead of classed here to enable passing the right type.

File: packages/shared/src/components/squads/moderation/SquadModerationItem.tsx [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+import { SquadModerationActions } from './SquadModerationActions';
+import { ProfileImageSize, ProfilePicture } from '../../ProfilePicture';
+import PostMetadata from '../../cards/common/PostMetadata';
+import { AlertPointerMessage } from '../../alert/common';
+import { SourceMemberRole, Squad } from '../../../graphql/sources';
+import { Button, ButtonSize, ButtonVariant } from '../../buttons/Button';
+import { capitalize } from '../../../lib/strings';
+import OptionsButton from '../../buttons/OptionsButton';
+import { TimerIcon } from '../../icons';
+import { WarningIcon } from '../../icons/Warning';
```
Comment (rebelchris) [inline packages/shared/src/components/squads/moderation/SquadModerationItem.tsx:22]: If you export warning from icons we can do 1 barrel import here.

File: packages/shared/src/components/squads/moderation/SquadModerationList.tsx (hunk 4 above)
Comment (rebelchris) [reply packages/shared/src/components/squads/moderation/SquadModerationList.tsx:25]: Probably good idea indeed. :) 
Just basic one based on the elements is fine

File: packages/shared/src/components/squads/moderation/SquadModerationList.tsx [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        </div>
+      );
+    }
+
+    if (squad.currentMember.role === SourceMemberRole.Member) {
+      return (
+        <SquadEmptyScreen
+          Icon={VIcon}
+          title="All done!"
+          description="All caught up! No posts are pending" // check with product what to show here
```
Comment (rebelchris) [inline packages/shared/src/components/squads/moderation/SquadModerationList.tsx:35]: Please add ticket/todo here :)

Comment (rebelchris) [reply packages/shared/src/components/squads/moderation/SquadModerationList.tsx:35]: Also this is for the member, not the admin?

File: packages/shared/src/graphql/squads.ts (hunk 3 above)
Comment (rebelchris) [reply packages/shared/src/graphql/squads.ts:611]: Just add a Todo maybe so we don't forget

File: packages/shared/src/hooks/squads/useSquadPendingPosts.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  data: SourcePostModeration[];
+  isFetched: boolean;
+}
+
+export const useSquadPendingPosts = (squadId: string): UseSquadPendingPosts => {
+  const { user } = useAuthContext();
+  const { data } = useQuery(
+    generateQueryKey(RequestKey.SquadPostRequests, user, squadId),
+    () =>
+      Promise.resolve([
```
Comment (rebelchris) [inline packages/shared/src/hooks/squads/useSquadPendingPosts.ts:20]: Probably also good to add Todo here :)

File: packages/shared/src/hooks/squads/useSquadPostModeration.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+} from '../../graphql/squads';
+import { useLazyModal } from '../useLazyModal';
+import { LazyModal } from '../../components/modals/common/types';
+import { usePrompt } from '../usePrompt';
+import { useToastNotification } from '../useToastNotification';
+
+export const rejectReasons: { value: PostModerationReason; label: string }[] = [
+  {
+    value: PostModerationReason.OffTopic,
+    label: 'Off-topic post unrelated to the squad’s purpose',
```
Comment (rebelchris) [inline packages/shared/src/hooks/squads/useSquadPostModeration.ts:17]: Please make sure we always write `Squad` capitalized :)

File: packages/webapp/pages/squads/[handle]/moderate.tsx [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  const router = useRouter();
+
+  useEffect(() => {
+    if (!isFetched) {
+      return;
+    }
+
+    if (squad?.moderationRequired) {
+      return;
+    }
```
Comment (rebelchris) [inline packages/webapp/pages/squads/[handle]/moderate.tsx:43]: This is intented? (not sure if it made more sense to show some copy to say it's disabled if your squad doesn't have this setting)

File: packages/webapp/pages/squads/[handle]/moderate.tsx [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+    router.push(squad ? squad.permalink : '/404');
+  }, [squad, isFetched, router]);
+
+  return (
+    <ManageSquadPageContainer>
+      <PageHeader className="border-b-0">
```
Comment (rebelchris) [inline packages/webapp/pages/squads/[handle]/moderate.tsx:50]: I see in many places we need to overwrite to 0 border bottom, wouldn't it make more sense to not have a border and then pass the one place you need it?

File: packages/webapp/pages/squads/[handle]/moderate.tsx [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+export async function getStaticPaths(): Promise<GetStaticPathsResult> {
+  return { paths: [], fallback: true };
+}
+
+interface SquadPageParams extends ParsedUrlQuery {
+  handle: string;
+}
+
+export function getStaticProps({
```
Comment (rebelchris) [inline packages/webapp/pages/squads/[handle]/moderate.tsx:77]: We could probably load the squad server side here?

**Review by rebelchris — CHANGES_REQUESTED** [review]
Some minor feedback, impressive work.
Just review based on code, didn't check online.

File: packages/webapp/pages/squads/[handle]/moderate.tsx (hunk 11 above)
Comment (ilasw) [reply packages/webapp/pages/squads/[handle]/moderate.tsx:43]: maybe we want to split the logic:
- if admin -> Copy that push for going to settings and change them;
- if user -> redirect to squad page;

wdyt?

File: packages/shared/src/components/layout/common.tsx (hunk 6 above)
Comment (sshanzel) [reply packages/shared/src/components/layout/common.tsx:172]: Right, let me update it. We should probably build a similar `classed` function to wrap the Typography. The result is kinda verbose for quick cases like this. For example, this would become at least a 5-liner code.

File: packages/shared/src/components/squads/moderation/SquadModerationList.tsx (hunk 8 above)
Comment (sshanzel) [reply packages/shared/src/components/squads/moderation/SquadModerationList.tsx:35]: There is also a page for members to see their pending/rejected posts.

File: packages/webapp/pages/squads/[handle]/moderate.tsx (hunk 11 above)
Comment (sshanzel) [reply packages/webapp/pages/squads/[handle]/moderate.tsx:43]: @ilasw sounds good to me 🚀

File: packages/webapp/pages/squads/[handle]/moderate.tsx (hunk 12 above)
Comment (sshanzel) [reply packages/webapp/pages/squads/[handle]/moderate.tsx:50]: This is used somewhere where the `border-b` is necessary. Also, that is the default behavior for headers, it just so happened here we reused it on top another header (tabs).

File: packages/webapp/pages/squads/[handle]/moderate.tsx (hunk 13 above)
Comment (sshanzel) [reply packages/webapp/pages/squads/[handle]/moderate.tsx:77]: Thought of that too intially, but what data will we load from there?

File: packages/webapp/pages/squads/[handle]/moderate.tsx (hunk 11 above)
Comment (sshanzel) [reply packages/webapp/pages/squads/[handle]/moderate.tsx:43]: They won't have a way to go to this page if it is disabled, hence redirection works for me.

File: packages/shared/src/hooks/squads/useSquadPendingPosts.ts (hunk 9 above)
Comment (sshanzel) [reply packages/shared/src/hooks/squads/useSquadPendingPosts.ts:21]: TODOs are now added.

File: packages/shared/src/hooks/squads/useSquadPostModeration.ts (hunk 10 above)
Comment (sshanzel) [reply packages/shared/src/hooks/squads/useSquadPostModeration.ts]: Right. Should be updated now.

File: packages/shared/src/components/squads/moderation/SquadModerationItem.tsx (hunk 7 above)
Comment (sshanzel) [reply packages/shared/src/components/squads/moderation/SquadModerationItem.tsx]: When do barrel imports work for us? I remember some discussion to avoid it as much as possible, is it not the case with the icons?

File: packages/shared/src/components/layout/common.tsx (hunk 6 above)
Comment (rebelchris) [reply packages/shared/src/components/layout/common.tsx]: Ah I was thinking it would still be a `common` element so you can simply only extend it's attributes?
(This is just top of mind maybe it doesn't work for reasons, happy to show you what I had in mind)

File: packages/shared/src/components/squads/moderation/SquadModerationItem.tsx (hunk 7 above)
Comment (rebelchris) [reply packages/shared/src/components/squads/moderation/SquadModerationItem.tsx]: With Icons it's normally safe as they don't export other functions that can cause infinite cycle.
For now let's keep using them for icons.

File: packages/webapp/pages/squads/[handle]/moderate.tsx (hunk 11 above)
Comment (rebelchris) [reply packages/webapp/pages/squads/[handle]/moderate.tsx:49]: Sounds good indeed.

File: packages/webapp/pages/squads/[handle]/moderate.tsx (hunk 12 above)
Comment (rebelchris) [reply packages/webapp/pages/squads/[handle]/moderate.tsx:59]: Ah ok, lets leave it as is then yes.

File: packages/webapp/pages/squads/[handle]/moderate.tsx (hunk 13 above)
Comment (rebelchris) [reply packages/webapp/pages/squads/[handle]/moderate.tsx:86]: You can choose what data you need basically.
Whatever you need to render the page should be there in that case.

File: packages/shared/src/components/squads/moderation/SquadModerationList.tsx (hunk 4 above)
Comment (sshanzel) [reply packages/shared/src/components/squads/moderation/SquadModerationList.tsx]: Should look like below 👀 

Screenshot 2024-10-09 at 8 37 57 PM

File: packages/webapp/pages/squads/[handle]/moderate.tsx (hunk 13 above)
Comment (sshanzel) [reply packages/webapp/pages/squads/[handle]/moderate.tsx]: Yep. Also moved the redirection parts on the server side 👀

File: packages/shared/src/components/squads/moderation/SquadModerationItem.tsx (hunk 7 above)
Comment (sshanzel) [reply packages/shared/src/components/squads/moderation/SquadModerationItem.tsx]: Yeah, that makes sense. Thank you!

File: packages/shared/src/components/layout/common.tsx (hunk 6 above)
Comment (sshanzel) [reply packages/shared/src/components/layout/common.tsx]: Oh, cool. Yeah, we can maybe check that tomorrow and see what we can come up with 👀

File: packages/shared/src/components/squads/moderation/SquadModerationList.tsx (hunk 8 above)
Comment (rebelchris) [reply packages/shared/src/components/squads/moderation/SquadModerationList.tsx:66]: Just add a TODO here please :)

**Review by rebelchris — APPROVED** [review]
One todo missed, otherwise good from my side so far.

File: packages/shared/src/graphql/squads.ts (hunk 3 above)
Comment (sshanzel) [reply packages/shared/src/graphql/squads.ts:636]: Should have the todo now 🚀

File: packages/shared/src/components/squads/moderation/SquadModerationList.tsx (hunk 8 above)
Comment (sshanzel) [reply packages/shared/src/components/squads/moderation/SquadModerationList.tsx:66]: Right. Forgot about this one. Let me create a ticket.

File: packages/shared/src/components/modals/common.tsx [hunk 14]
```diff
@@ -196,6 +200,7 @@ export const modals = {
   [LazyModal.ReportSource]: ReportSourceModal,
   [LazyModal.UserFollowersModal]: UserFollowersModal,
   [LazyModal.UserFollowingModal]: UserFollowingModal,
+  [LazyModal.Report]: ReportModal,
```
Comment (nensidosari) [inline packages/shared/src/components/modals/common.tsx:203]: Why is this called report? think it confuses things, I believe it's best to be called PostRejectModal

Comment (nensidosari) [inline packages/shared/src/components/modals/common.tsx:203]: Why is this called report? think it confuses things, I believe it's best to be called PostRejectModal

File: packages/shared/src/components/modals/common/types.ts [hunk 15]
```diff
@@ -36,6 +36,7 @@ export enum LazyModal {
   ReadingHistory = 'readingHistory',
   SquadPromotion = 'squadPromotion',
   CreateSharedPost = 'createSharedPost',
+  Report = 'report',
```
Comment (nensidosari) [inline packages/shared/src/components/modals/common/types.ts:39]: I think it makes more sense to say rejectPost here

File: packages/shared/src/components/modals/report/ReportModal.tsx [hunk 16]
```diff
@@ -17,15 +19,17 @@ interface Props extends ModalProps {
 
 export const OTHER_KEY = 'OTHER';
 
-export function ReportModal({
+export function ReportModal<
+  T extends ReportReason | PostModerationReason = ReportReason,
```
Comment (nensidosari) [inline packages/shared/src/components/modals/report/ReportModal.tsx:23]: As comment above, better named reject then report

File: packages/shared/src/components/squads/moderation/SquadEmptyScreen.tsx [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  TypographyType,
+} from '../../typography/Typography';
+
+interface SquadEmptyScreenProps {
+  Icon: FC<IconProps>;
+  title: string;
+  description: string;
+}
+
+export function SquadEmptyScreen({
```
Comment (nensidosari) [inline packages/shared/src/components/squads/moderation/SquadEmptyScreen.tsx:15]: Would suggest to remove Squad from naming, not necessary, it's already within squads directory, and same comment for the rest

**Review by nensidosari — COMMENTED** [review]
Looks great, just some post merge feedback 😅 
Screenshot 2024-10-10 at 11 44 21 AM

File: packages/shared/src/components/modals/common.tsx (hunk 14 above)
Comment (sshanzel) [reply packages/shared/src/components/modals/common.tsx:203]: This was just a reused pre-existing base modal we have. Maybe I can rename this to `ReasonSelectionModal` to be more clear.

File: packages/shared/src/components/squads/moderation/SquadEmptyScreen.tsx (hunk 17 above)
Comment (sshanzel) [reply packages/shared/src/components/squads/moderation/SquadEmptyScreen.tsx:15]: Yeah, but normally, the main function's name should be the same as with the file and we do named export, removing the Squad on the name would cause some naming issues down the line.

File: packages/shared/src/components/modals/report/ReportModal.tsx (hunk 16 above)
Comment (sshanzel) [reply packages/shared/src/components/modals/report/ReportModal.tsx:23]: The `ReportReason` exists for some other reason (pun intended 🤣). This is used for the post/comment/source report modal.

For the case of Squad approval, we use `PostModerationReason`.

File: packages/shared/src/components/modals/common.tsx (hunk 14 above)
Comment (sshanzel) [reply packages/shared/src/components/modals/common.tsx:203]: Would love to hear what you think here: https://github.com/dailydotdev/apps/pull/3656

**Comment by sshanzel** [comment]
@ilasw sorry, forgot to wait for your review, let me know if you want anything to be changed, and I shall raise another PR 🙏
